### ISSUE TITLE
Force Padding to a fixed size

### ISSUE THREAD

**Issue by prochri** [cloudflare/quiche#1494]
I am using quiche in a privacy setting and want to hide the packet size by transferring only short header packets with the similar sizes.
Ideally, this would allow me to provide a set of allowed QUIC packet sizes and quiche would automatically pad the payload to the next allowed size.

Is this possible with the current implementation? I did not find any method to add such a padding.

If not, how could this be implemented?

**Comment by LPardue** [comment]
FWIW this form of padding might not achieve the privacy goals you have in mind.

We don't offer this capability but it is possible to add it to quiche as a custom patch, you'd want to modify `[send_single](https://github.com/cloudflare/quiche/blob/master/quiche/src/lib.rs#L3248)` to do some accounting ad frames are added to a packet and once you get to the end, add paading up to the size you want.

**Comment by ghedo** [comment]
You can also pad the UDP datagrams in your application directly. That is, once quiche writes a packet into the output buffer, an application can fill the remainder of the UDP datagram with random bytes up to whatever padded size you want. Conforming QUIC implementations should then just ignore the additional bytes that are not part of an actual encrypted QUIC packet when receiving UDP datagrams.

Not only does this not require changes to quiche (as it can be implemented in the application), but it is also much more efficient than having to encrypt a whole bunch of padding bytes. It is also the strategy that quiche (see https://github.com/cloudflare/quiche/blob/master/quiche/src/lib.rs#L3225-L3234) and other QUIC implementations use for padding in most cases, so it's already proven to work in practice.

**Comment by prochri** [comment]
Thank you for your responses!
I explored this topic as part of my Master's thesis, which I have finished now, without implementing padding due to time limitations.

I am fully aware that padding, especially when there are multiple size options, can not provide the privacy goals alone.

Regarding padding in the application as suggested by @ghedo, the given example applies only to initial packets, which contain the payload length as header field. Short header packets do not have such a field. Instead, the application must assume, that the remaining packet belongs to the same packet (see [these lines in recv_single](https://github.com/cloudflare/quiche/blob/ab2b44fa9b8b0f82d6604cb5693447deaf07dfdc/quiche/src/lib.rs#L2421-L2425C10)).

While I am no longer that involved with the topic, I do not thinkthat AEAD is able to deal with this unknown ciphertext size when decrypting. A short test (adding padding in the example http3-server) seems to confirm this.
